### DISCUSSION TITLE
(v8) Use `onColumnFiltersChange` to set URL search params when column filters change creates a tree mismatch and breaks the render cycle

### DISCUSSION THREAD

**Discussion by bennettdams** [TanStack/table#5002] [Q&A]
Hey!

I would like to update the URL search params whenever a column filter is changing.

When one uses the `onColumnFiltersChange` function, it mimics how a `useState`'s setter works: You either set the state directly or you receive an updater function which also allows access to the previous state.

Here's what I initially tried to do:

```ts
  const table = useReactTable({
    ...
    onColumnFiltersChange: (filtersOrUpdaterFn) => {
      if (typeof filtersOrUpdaterFn !== "function") {
        setColumnFilters(filtersOrUpdaterFn)

        // this works fine, but in my case, this variant of the callback is never used
        setSearchParam(filtersOrUpdaterFn)
      } else {
        setColumnFilters((old) => {
          const newFilters = filtersOrUpdaterFn(old)

          // 💣this does not work
          setSearchParam(newFilters)

          return newFilters
        })
      }
    },
    ...
  })
```
...but as you maybe know, updating INSIDE the setter's update function is not allowed.
So in my example, `setSearchParam` creates a tree mismatch and breaks the render cycle and the following error is produced:

![image](https://github.com/TanStack/table/assets/29319414/8e737a21-bb6e-4a13-b5ff-92b73a2ed934)

### Workaround

Right now, I keep the column filters in a `useState` and only update the search params in a `useEffect` like so, which is obviously not ideal:

```ts
  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>(
    !paramValue ? initialColumnFilters : JSON.parse(paramValue),
  )
  
  ...
  onColumnFiltersChange: setColumnFilters,
  ...

  useEffect(() => {
    /**
     * Unfortunately we can't set the search params in the `onColumnFiltersChange` function, as it only uses the setters update function.
     * Doing it inside the setter's callback would create a tree mismatch with such an error:
     * "Cannot update a component (`Router`) while rendering a different component (`DataTable`)......"
     */
    if (isColumnsFiltered) {
      setSearchParam(columnFilters)
    } else {
      deleteSearchParam()
    }
  }, [columnFilters, setSearchParam, isColumnsFiltered, deleteSearchParam])
```

---

What is the canonical way to do "other" stuff in the `onColumnFiltersChange` callback?
Can one somehow force the `onColumnFiltersChange` to never use the "updater" style of the setter and only use the variant that provides the state directly?

**Comment by muuhoffman** [comment]
For what it's worth, your answer helped me solve my same issue, and I'm not dealing with any type errors. However, I'm still having to do the `typeof updater === "function"` check.

```typescript
  const [{sortBy, filterBy}, setSearchParams] = useSearchParams(); // my custom search params hook that returns the string or default of the search params

  const sorting = useMemo(() => {
    try {
      return JSON.parse(sortBy);
    } catch (e) {
      return [];
    }
  }, [sortBy]);
  const setSorting: OnChangeFn<SortingState> = useCallback(
    (updater) => {
      const newSorting =
        typeof updater === "function" ? updater(sorting) : updater;
      setSearchParams({
        sortBy: JSON.stringify(newSorting),
      });
    },
    [setSearchParams, sorting]
  );
  const columnFilters = useMemo(() => {
    try {
      return JSON.parse(filterBy);
    } catch (e) {
      return [];
    }
  }, [filterBy]);
  const setColumnFilters: OnChangeFn<ColumnFiltersState> = useCallback(
    (updater) => {
      const newColumnFilters =
        typeof updater === "function" ? updater(columnFilters) : updater;
      setSearchParams({
        filterBy: JSON.stringify(newColumnFilters),
      });
    },
    [setSearchParams, columnFilters]
  );

  const table = useReactTable({
    columns: MY_COLUMNS,
    data: tableData,
    getRowId: (row) => row.id,
    getCoreRowModel: getCoreRowModel(),
    // sorting
    getSortedRowModel: getSortedRowModel(),
    onSortingChange: setSorting,
    // filtering
    getFilteredRowModel: getFilteredRowModel(),
    getFacetedRowModel: getFacetedRowModel(),
    getFacetedUniqueValues: getFacetedUniqueValues(),
    onColumnFiltersChange: setColumnFilters,
    state: {
      sorting,
      columnFilters,
    },
  });
  ```

**Comment by mike-weiner** [comment]
I'm looking at adding functionality to reflect the state of my table's filtering and sorting into URL query parameters here: https://github.com/Minnesota-Computer-Club/MCC-Website-v2/blob/main/components/Table/index.js

What is the recommended approach for doing something like that? Is it still something like the above?

**Reply by muuhoffman** [reply]
My approach above has been working well for me, I would recommend it if you're looking to store the sort/filter state of your table across page refreshes.

**Reply by mike-weiner** [reply]
Thanks, @muuhoffman. Do you have any code available where you're using the above?

**Comment by xixixao** [comment]
@tannerlinsley I think this API sucks. It should be `ColumnFiltersState => void`. Otherwise it's hard to save the state elsewhere besides useState (requiring going through a ref to avoid infinite render cycle).

**Reply by xixixao** [reply]
Also found out that this lib is incompatible with React Compiler. Ended up rolling my own code (well, AI, ofc...). Works well enough for my use case. Doesn't feel `@tanstack/table` is on the same level as `@tanstack/query`.

**Comment by yudin-s** [comment]
You can't force TanStack Table to only call the next-value form. The `OnChangeFn<T>` contract intentionally mirrors React state setters, so consumers need to handle both `T` and `(old: T) => T`.

The canonical pattern is to make the URL the controlled source, compute the current table state from it, and implement the updater yourself outside a React state setter:

```ts
const columnFilters = useMemo(() => parseFilters(searchParams), [searchParams])

const onColumnFiltersChange: OnChangeFn<ColumnFiltersState> = useCallback(
  (updater) => {
    const next = typeof updater === 'function'
      ? updater(columnFilters)
      : updater

    const nextSearch = serializeFilters(next)

    if (nextSearch !== searchParams.get('filters')) {
      setSearchParams({ filters: nextSearch })
    }
  },
  [columnFilters, searchParams, setSearchParams],
)
```

Then pass both:

```ts
useReactTable({
  state: { columnFilters },
  onColumnFiltersChange,
})
```

The key is: do not call `setSearchParams` from inside another state setter callback. Compute `next` synchronously from the current controlled value, then update the URL once. Also make serialization deterministic so you do not create a URL/table feedback loop.
